lib/jekyll/converters/markdown/kramdown_parser.rb

```diff
@@ -1,33 +1,84 @@
+# Frozen-string-literal: true
+# Encoding: utf-8
+
 module Jekyll
   module Converters
     class Markdown
       class KramdownParser
+        CODERAY_DEFAULTS = {
+          "bold_every" => 10,
+          "css" => "style",
+          "line_numbers" => "inline",
+          "line_number_start" => 1,
+          "wrap" => "div",
+          "tab_width" => 4
+        }
+
         def initialize(config)
-          require 'kramdown'
-          @config = config
-          # If kramdown supported highlighter enabled, use that
-          highlighter = @config['highlighter']
-          if highlighter == 'rouge' || highlighter == 'coderay'
-            @config['kramdown']['syntax_highlighter'] ||= highlighter
-          end
-        rescue LoadError
-          STDERR.puts 'You are missing a library required for Markdown. Please run:'
-          STDERR.puts '  $ [sudo] gem install kramdown'
-          raise Errors::FatalException.new("Missing dependency: kramdown")
+          Jekyll::External.require_with_graceful_fail "kramdown"
+          @main_fallback_highlighter = config["highlighter"] || "rogue"
+          @config = config["kramdown"] || {}
+          setup
+        end
+
+        # Setup and normalize the configuration:
+        #   * Create Kramdown if it doesn't exist.
+        #   * Set syntax_highlighter, detecting enable_coderay and merging highlighter if none.
+        #   * Merge kramdown[coderay] into syntax_highlighter_opts stripping coderay_.
+        #   * Make sure `syntax_highlighter_opts` exists.
+
+        def setup
+          @config["syntax_highlighter"] ||= highlighter
+          @config["syntax_highlighter_opts"] ||= {}
+          @config["coderay"] ||= {} # XXX: Legacy.
+          modernize_coderay_config
         end
 
         def convert(content)
-          # Check for use of coderay
-          if @config['kramdown']['enable_coderay']
-            %w[wrap line_numbers line_numbers_start tab_width bold_every css default_lang].each do |opt|
-              key = "coderay_#{opt}"
-              @config['kramdown'][key] = @config['kramdown']['coderay'][key] unless @config['kramdown'].key?(key)
+          Kramdown::Document.new(content, @config).to_html
+        end
+
+        # config[kramdown][syntax_higlighter] > config[kramdown][enable_coderay] > config[highlighter]
+        # Where `enable_coderay` is now deprecated because Kramdown
+        # supports Rouge now too.
+
+        private
+        def highlighter
+          @highlighter ||= begin
+            if out = @config["syntax_highlighter"] then out
+            elsif @config.has_key?("enable_coderay") && @config["enable_coderay"]
```

Comment:
I believe `has_key?` has been deprecated in favour of `key?`. Would you mind updating here and below?